src/Connect.jl

```diff
@@ -0,0 +1,90 @@
+Base.flush(sock::ClickHouseSock) = Base.flush(sock.io)
+
+function Base.close(sock::ClickHouseSock)
+    @guarded sock begin
+        !isnothing(sock.io) && Base.close(sock.io)
+        sock.io = nothing
+        sock.busy = false
+    end
+end
+
+function connect!(sock::ClickHouseSock; force = false)
+    force && is_connected(sock) && close(sock)
+    is_connected(sock) && return sock
+
+    try
+        set_busy!(sock, true)
+        tcp = Sockets.TCPSocket()
+        Base.buffer_writes(tcp, sock.settings.send_buffer_size)
+        Sockets.connect!(tcp, sock.settings.host, sock.settings.port)
+
+        timeout = Ref{Bool}(false)
+        @async begin
+            sleep(sock.settings.connection_timeout)
+            if tcp.status == Sockets.StatusConnecting
+                timeout[] = true
+                tcp.status = Base.StatusClosing
+                ccall(:jl_forceclose_uv, Nothing, (Ptr{Nothing},), tcp.handle)
+            end
+        end
+        try
+            Sockets.wait_connected(tcp)
+        catch e
+            if timeout[]
+                error("Connection timeout")
+            end
+            rethrow(e)
+        end
+
+        @guarded sock sock.io = tcp
+
+        write_packet(sock, ClientHello(
+            CLIENT_NAME,
+            DBMS_VER_MAJOR,
+            DBMS_VER_MINOR,
+            DBMS_VER_REV,
+            @guarded(sock, sock.settings.database),
+            @guarded(sock, sock.settings.username),
+            @guarded(sock, sock.settings.password)
+        ))
+        info = read_packet(sock, ServerCodes)::ServerInfo
+        @guarded sock begin
+            sock.server_name = isempty(info.server_display_name) ?
+                            info.server_name : info.server_display_name
+            sock.server_rev = UInt64(info.server_rev)
+            sock.server_timezone = info.server_timezone
+            sock.busy = false
+        end
```

Comment:
Acquiring and releasing a lock is a somewhat expensive operations. Also, the code above uses sock without synchronizing. Can't we just wrap the whole function content in one big guard?

Reply:
Partially done at df9f9ed . Now two protected blocks:
```julia
hello = @guarded sock begin
            sock.io = tcp
            ClientHello(
                CLIENT_NAME,
                DBMS_VER_MAJOR,
                DBMS_VER_MINOR,
                DBMS_VER_REV,
                sock.settings.database,
                sock.settings.username,
                sock.settings.password
            )

        end

        write_packet(sock, hello)
        info = read_packet(sock, ServerCodes)::ServerInfo
        @guarded sock begin
            sock.server_name = isempty(info.server_display_name) ?
                            info.server_name : info.server_display_name
            sock.server_rev = UInt64(info.server_rev)
            sock.server_timezone = info.server_timezone
            sock.busy = false
        end
```
Reading and writing cannot be under lock, because these are potentially very slow operations. My main motivation for introducing the lock is to allow users to create something like a pool of connections. If checking "whether this connection is free" will stay in the lock waiting for communication with the server in another thread, then this will kill the whole point. On the other hand, `connect ' itself is not an operation in which performance is so important

Reply:
My idea was as follows - lock protects changing statuses (I also plan to add the status of the current request aka ProgressInfo to the socket). Busy status protects the connection from simultaneous IO

Reply:
Alright, I see, makes sense. I was about to say that we could just use [atomics](https://docs.julialang.org/en/v1/base/multi-threading/#Atomic-operations) that are pretty close to free of overhead on X86 for that, but it looks like they are still unstable/"beta" in Julia, so doing things as you are right now is likely for the best.

Reply:
I did not think in the direction of atomics first of all because I planned to add complex types of current progress in addition to simple statuses like busy. Yes, and the same io `{Nothing, IO}` so-so to do atomic. In addition, the presence of `ReinterpretLock` field is a convenient opportunity for the user to add their own protected behavior associated with the socket. 
If you are talking about Atomics hinted at `SpinLock`, then I am quite critical of it. In the loaded sections of the code very carefully-Yes, but I always weigh their need well before applying

Reply:
I was primarily thinking of using atomics for the progress tracking and busy status. But yeah, it's fine as is -- not need to complicate this unless it actually turns out to be a bottleneck.